runtime/compiler/runtime/JITServerAOTDeserializer.cpp

```diff
@@ -898,3 +901,600 @@ JITServerLocalSCCAOTDeserializer::updateSCCOffsets(SerializedAOTMethod *method,
+   getRAMClassChain(comp, firstClass, ramClassChain, ramClassChainLength);
+
+   // Check that it has the expected length
+   if (record->list().length() != ramClassChainLength)
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+            "ERROR: Class chain length mismatch for class %.*s ID %zu: %zu != %zu",
+            ROMCLASS_NAME(firstClass->romClass), record->list().ids()[0], ramClassChainLength, record->list().length()
+         );
+      return false;
+      }
+
+   // Validate each class in the server's chain (which should all be cached by now)
+   for (size_t i = 0; i < ramClassChainLength; ++i)
+      {
+      auto ramClass = findInMap(_classIdMap, record->list().ids()[i], getClassMonitor(), comp, wasReset)._ramClass;
+      if (!ramClass)
+         {
+         if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+               "ERROR: Class %.*s ID %zu mismatch or invalidation in class chain ID %zu for class %.*s ID %zu",
+               ROMCLASS_NAME(ramClassChain[i]->romClass), record->list().ids()[i], record->id(),
+               ROMCLASS_NAME(firstClass->romClass), record->list().ids()[0]
+            );
+         return false;
+         }
+
+      if (ramClass != ramClassChain[i])
+         {
+         if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+               "ERROR: Class %.*s mismatch in class chain ID %zu for class %.*s ID %zu",
+               ROMCLASS_NAME(ramClassChain[i]->romClass), record->id(),
+               ROMCLASS_NAME(firstClass->romClass), record->list().ids()[0]
+            );
+         return false;
+         }
+      }
+
+   // Chain validation is now complete
+   size_t chainSize = (record->list().length() + 1) * sizeof(uintptr_t); // offset chains are expected to store their size in bytes in their first entry
+   auto deserializerChain = (uintptr_t *)TR::Compiler->persistentGlobalMemory()->allocatePersistentMemory(chainSize);
+   if (!deserializerChain)
+      throw std::bad_alloc();
+
+   deserializerChain[0] = chainSize;
+   uintptr_t *deserializerChainData = deserializerChain + 1;
+   for (size_t i = 0; i < record->list().length(); ++i)
+      deserializerChainData[i] = encodeClassOffset(record->list().ids()[i]);
+   addToChainMap(_classChainMap, it, record->id(), deserializerChain);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+         "Cached class chain record ID %zu -> { %p } for class %.*s ID %zu",
+         record->id(), firstClass, ROMCLASS_NAME(firstClass->romClass), record->list().ids()[0]
+      );
+   return true;
+   }
+
+// See revalidateClassChain()
+bool
+JITServerNoSCCAOTDeserializer::revalidateWellKnownClasses(uintptr_t *wellKnownClassesChain, TR::Compilation *comp, bool &wasReset)
```

Comment:
In a local AOT, the WKC validation happens before normal relocation, because it's one of the first entries in the binary template buffer. In the case of serialization on the server side, do we ensure that the WKC Serialization Record is created after we've already serialized all the class chain records?

I ask because this function assumes that all class chains associated with this compiled method has already been deserialized. I assume this is true even for `JITServerLocalSCCAOTDeserializer` though right?